adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/uap/PlatformInformation.cs

```diff
@@ -53,21 +56,69 @@ public override string GetEnvironmentVariable(string variable)
         }
 
+        /// <summary>
+        /// Get a principal name that can be used for WIA
+        /// </summary>
+        /// <remarks>
+        /// Win10 allows several identities to be logged in at once; 
+        /// select the first principal name that can be used
+        /// </remarks>
+        /// <returns></returns>
         public override async Task<string> GetUserPrincipalNameAsync()
         {
-            if (!UserInformation.NameAccessAllowed)
+            IReadOnlyList<User> users = await User.FindAllAsync();
+            if (users == null || !users.Any())
             {
-                throw new AdalException(AdalErrorEx.CannotAccessUserInformation, AdalErrorMessageEx.CannotAccessUserInformation);
+                throw new AdalException(
+                 AdalErrorEx.CannotAccessUserInformationOrUserNotDomainJoined,
+                 CoreErrorMessages.UapCannotFindDomainUser);
             }
 
-            try
+            var getUserDetailTasks = users.Select(async u =>
             {
-                return await UserInformation.GetPrincipalNameAsync().AsTask().ConfigureAwait(false);
+                object domainObj = await u.GetPropertyAsync(KnownUserProperties.DomainName);
+                string domainString = domainObj?.ToString();
+
+                object principalObject = await u.GetPropertyAsync(KnownUserProperties.PrincipalName);
+                string principalNameString = principalObject?.ToString();
+
+                return new { Domain = domainString, PrincipalName = principalNameString };
+            }).ToList();
+
+            var userDetails = await Task.WhenAll(getUserDetailTasks).ConfigureAwait(false);
+
+            // try to get a user that has both domain name and upn
+            var userDetailWithDomainAndPn = userDetails.FirstOrDefault(
+                d => !String.IsNullOrWhiteSpace(d.Domain) &&
+                !String.IsNullOrWhiteSpace(d.PrincipalName));
+
+            if (userDetailWithDomainAndPn != null)
+            {
+                return userDetailWithDomainAndPn.PrincipalName;
             }
-            catch (UnauthorizedAccessException ex)
+
+            // try to get a user that at least has upn
+            var userDetailWithPn = userDetails.FirstOrDefault(
+              d => !String.IsNullOrWhiteSpace(d.PrincipalName));
+
+            if (userDetailWithPn != null)
+            {
+                return userDetailWithPn.PrincipalName;
+            }
+
+            // user has domain name, but no upn -> missing Enterprise Auth capability
+            if (userDetails.Any(d => !String.IsNullOrWhiteSpace(d.Domain)))
             {
-                throw new AdalException(AdalErrorEx.UnauthorizedUserInformationAccess, AdalErrorMessageEx.UnauthorizedUserInformationAccess, ex);
+                throw new AdalException(
+                 AdalErrorEx.CannotAccessUserInformationOrUserNotDomainJoined,
+                 CoreErrorMessages.UapCannotFindUpn);
```

Comment:
"Cannot find the principal name for the user, but found a domain the name. Possible cause: the application does not request the Enterprise Authentication capability.";